File: src/cpp/query_engine/QueryAnswer.cc
#include "QueryAnswer.h"
#include "Utils.h"
#include <cmath>
#include <cstring>
#include <iostream>

using namespace query_engine;
using namespace commons;

// -------------------------------------------------------------------------------------------------
// Assignment

Assignment::Assignment() {
    this->size = 0;
}

Assignment::~Assignment() {
}

bool Assignment::assign(const char *label, const char *value) {
    for (unsigned int i = 0; i < this->size; i++) {
        // if label is already present, return true iff its value is the same
        if (strncmp(label, this->labels[i], MAX_VARIABLE_NAME_SIZE) == 0) {
            return (strncmp(value, this->values[i], HANDLE_HASH_SIZE) == 0);
        }
    }
    // label is not present, so makes the assignment and return true
    labels[this->size] = label;
    values[this->size] = value;
    this->size++;
    if (this->size == MAX_NUMBER_OF_VARIABLES_IN_QUERY) {
        Utils::error(
            "Assignment size exceeds the maximal number of allowed variables in a query: " +
            std::to_string(MAX_NUMBER_OF_VARIABLES_IN_QUERY));
    }
    return true;
}

bool Assignment::is_compatible(const Assignment &other) {
    for (unsigned int i = 0; i < this->size; i++) {
        for (unsigned int j = 0; j < other.size; j++) {
            if ((strncmp(this->labels[i], other.labels[j], MAX_VARIABLE_NAME_SIZE) == 0) &&
                (strncmp(this->values[i], other.values[j], HANDLE_HASH_SIZE) != 0)) {
                   return false;
            }
        }
    }
    return true;
}

void Assignment::copy_from(const Assignment &other) {
    this->size = other.size;
    unsigned int num_bytes = this->size * sizeof(char *);
    memcpy((void *) this->labels, (const void *) other.labels, num_bytes);
    memcpy((void *) this->values, (const void *) other.values, num_bytes);
}

void Assignment::add_assignments(const Assignment &other) {
    bool already_contains;
    for (unsigned int j = 0; j < other.size; j++) {
        already_contains = false;
        for (unsigned int i = 0; i < this->size; i++) {
            if (strncmp(this->labels[i], other.labels[j], MAX_VARIABLE_NAME_SIZE) == 0) {
                already_contains = true;
                break;
            }
        }
        if (! already_contains) {
            this->labels[this->size] = other.labels[j];
            this->values[this->size] = other.values[j];
            this->size++;
        }
    }
}

const char *Assignment::get(const char *label) {
    for (unsigned int i = 0; i < this->size; i++) {
        if (strncmp(label, this->labels[i], MAX_VARIABLE_NAME_SIZE) == 0) {
            return this->values[i];
        }
    }
    return NULL;
}

unsigned int Assignment::variable_count() {
    return this->size;
}

string Assignment::to_string() {
    string answer = "{";
    for (unsigned int i = 0; i < this->size; i++) {
        answer += "(" + string(this->labels[i]) + ": " + string(this->values[i]) + ")";
        if (i != (this->size - 1)) {
            answer += ", ";
        }
    }
    answer += "}";
    return answer;
}

// -------------------------------------------------------------------------------------------------
// QueryAnswer


QueryAnswer::QueryAnswer() : QueryAnswer(0.0) {
}

QueryAnswer::QueryAnswer(double importance) {
    this->importance = importance;
    this->handles_size = 0;
}

QueryAnswer::QueryAnswer(const char *handle, double importance) {
    this->importance = importance;
    this->handles[0] = handle;
    this->handles_size = 1;
}

QueryAnswer::~QueryAnswer() {
}

void QueryAnswer::add_handle(const char *handle) {
    this->handles[this->handles_size++] = handle;
}

QueryAnswer *QueryAnswer::copy(QueryAnswer *base) { // Static method
    QueryAnswer *copy = new QueryAnswer(base->importance);
    copy->assignment.copy_from(base->assignment);
    copy->handles_size = base->handles_size;
    memcpy(
        (void *) copy->handles, 
        (const void *) base->handles, 
        base->handles_size * sizeof(char *));
    return copy;
}

bool QueryAnswer::merge(QueryAnswer *other, bool merge_handles) {
    if (this->assignment.is_compatible(other->assignment)) {
        this->assignment.add_assignments(other->assignment);
        bool already_exist;
        if (merge_handles) {
            this->importance = fmax(this->importance, other->importance);
            for (unsigned int j = 0; j < other->handles_size; j++) {
                already_exist = false;
                for (unsigned int i = 0; i < this->handles_size; i++) {
                    if (strncmp(this->handles[i], other->handles[j], HANDLE_HASH_SIZE) == 0) {
                        already_exist = true;
                        break;
                    }
                }
                if (! already_exist) {
                    this->handles[this->handles_size++] = other->handles[j];
                }
            }
        }
        return true;
    } else {
        return false;
    }
}

string QueryAnswer::to_string() {
    string answer = "QueryAnswer<" + std::to_string(this->handles_size) + ",";
    answer += std::to_string(this->assignment.variable_count()) + "> [";
    for (unsigned int i = 0; i < this->handles_size; i++) {
        answer += string(this->handles[i]);
        if (i != (this->handles_size - 1)) {
            answer += ", ";
        }
    }
    answer += "] " + this->assignment.to_string() + " " + std::to_string(this->importance);
    //answer += "] " + this->assignment.to_string();
    return answer;
}

const string &QueryAnswer::tokenize() {
    // char_count is computed to be slightly larger than actually required by assuming
    // e.g. 3 digits to represent sizes
    char importance_buffer[13];
    sprintf(importance_buffer, "%.10f", this->importance);
    unsigned int char_count = 
        13 // importance with 10 decimals + space
        + 4 // (up to 3 digits) to represent this->handles_size + space
        + this->handles_size * (HANDLE_HASH_SIZE + 1) // handles + spaces
        + 4 // (up to 3 digits) to represent this->assignment.size + space
        + this->assignment.size * (MAX_VARIABLE_NAME_SIZE + HANDLE_HASH_SIZE + 2); // label<space>handle<space>

    this->token_representation.clear();
    this->token_representation.reserve(char_count);
    string space = " ";
    this->token_representation += importance_buffer;
    this->token_representation += space;
    this->token_representation += std::to_string(this->handles_size);
    this->token_representation += space;
    for (unsigned int i = 0; i < this->handles_size; i++) {
        this->token_representation += handles[i];
        this->token_representation += space;
    }
    this->token_representation += std::to_string(this->assignment.size);
    this->token_representation += space;
    for (unsigned int i = 0; i < this->assignment.size; i++) {
        this->token_representation += this->assignment.labels[i];
        this->token_representation += space;
        this->token_representation += this->assignment.values[i];
        this->token_representation += space;
    }

    return this->token_representation;
}

static inline void read_token(
    const char *token_string, 
    unsigned int &cursor, 
    char *token, 
    unsigned int token_size) {

    unsigned int cursor_token = 0;
    while (token_string[cursor] != ' ') {
        if ((cursor_token == token_size) || (token_string[cursor] == '\0')) {
            Utils::error("Invalid token string");
        }
        token[cursor_token++] = token_string[cursor++];
    }
    token[cursor_token] = '\0';
    cursor++;
}

void QueryAnswer::untokenize(const string &tokens) {

    const char *token_string = tokens.c_str();
    char number[4];
    char importance[13];
    char handle[HANDLE_HASH_SIZE];
    char label[MAX_VARIABLE_NAME_SIZE];

    unsigned int cursor = 0;

    read_token(token_string, cursor, importance, 13);
    this->importance = std::stod(importance);

    read_token(token_string, cursor, number, 4);
    this->handles_size = (unsigned int) std::stoi(number);
    if (this->handles_size > MAX_NUMBER_OF_OPERATION_CLAUSES) {
        Utils::error("Invalid handles_size: " + std::to_string(this->handles_size) + " untokenizing QueryAnswer");
    }

    for (unsigned int i = 0; i < this->handles_size; i++) {
        read_token(token_string, cursor, handle, HANDLE_HASH_SIZE);
        this->handles[i] = strdup(handle);
    }

    read_token(token_string, cursor, number, 4);
    this->assignment.size = (unsigned int) std::stoi(number);

    if (this->assignment.size > MAX_NUMBER_OF_VARIABLES_IN_QUERY) {
        Utils::error("Invalid number of assignments: " + std::to_string(this->assignment.size) + " untokenizing QueryAnswer");
    }

    for (unsigned int i = 0; i < this->assignment.size; i++) {
        read_token(token_string, cursor, label, MAX_VARIABLE_NAME_SIZE);
        read_token(token_string, cursor, handle, HANDLE_HASH_SIZE);
        this->assignment.labels[i] = strdup(label);
        this->assignment.values[i] = strdup(handle);
    }

    if (token_string[cursor] != '\0') {
        Utils::error("Invalid token string - invalid text after QueryAnswer definition");
    }
}

File: 3rd_party_slots/python_client/hyperon_das/query_answer.py
from hyperon_das.logger import log


class Assignment:
    """
    Represents a set of variable assignments (label -> value).
    """

    def __init__(self) -> None:
        self._mapping = {}
        self.max_size = 100

    def assign(self, label: str, value: str) -> bool:
        """
        Adds a new assignment or verifies an existing one.

        If `label` already exists, returns True only if its existing value matches `value`.
        Otherwise, adds the (label, value) pair.

        Args:
            label (str): The variable label.
            value (str): The value to assign to the label.

        Returns:
            bool: True if the assignment was added or matches the existing value.

        Raises:
            ValueError: If the assignment size exceeds `max_size`.
        """
        if label in self._mapping:
            return self._mapping[label] == value
        else:
            if len(self._mapping) >= self.max_size:
                raise ValueError(f"Assignment size exceeds the maximal number of allowed variables in a query: {self.max_size}")
        self._mapping[label] = value
        return True

    def get(self, label: str) -> str | None:
        """
        Retrieves the value assigned to a label.

        Args:
            label (str): The variable label to look up.

        Returns:
            str: The assigned value if present, otherwise None.
        """
        return self._mapping.get(label)

    def is_compatible(self, other: "Assignment") -> bool:
        """
        Checks compatibility with another assignment.

        Two assignments are compatible if they never assign the same label to two different values.

        Args:
            other (Assignment): The other assignment to compare against.

        Returns:
            bool: True if compatible, False otherwise.
        """
        for label, value in self._mapping.items():
            if label in other._mapping and other._mapping[label] != value:
                return False
        return True

    def copy_from(self, other: "Assignment") -> None:
        """
        Replaces this assignment's contents with a copy of another's contents.

        Args:
            other (Assignment): The assignment to copy from.
        """
        self._mapping = other._mapping.copy()

    def add_assignments(self, other: "Assignment") -> None:
        """
        Adds assignments from another assignment that are not already present.

        Args:
            other (Assignment): The assignment whose entries to add.
        """
        for label, value in other._mapping.items():
            if label not in self._mapping:
                self._mapping[label] = value

    def variable_count(self) -> int:
        """
        Return the number of assigned variables.
        """
        return len(self._mapping)

    def to_string(self) -> str:
        """
        Return a string of the form: {(label1: value1), (label2: value2), ...}
        """
        pairs = [f"({label}: {value})" for label, value in self._mapping.items()]
        return "{" + ", ".join(pairs) + "}"

    def __eq__(self, other: object) -> bool:
        """
        Checks equality with another Assignment.

        Equality means same set of labels and values, order-independent.

        Args:
            other (object): The object to compare against.

        Returns:
            bool: True if `other` is Assignment with identical mappings.
        """
        if not isinstance(other, Assignment):
            return False
        return self._mapping == other._mapping


class QueryAnswer:
    """
    Encapsulates a set of handles, an importance score, and an associated Assignment.
    Provides methods for merging, tokenizing, and reconstructing.
    """

    def __init__(self, handle: str | None = None, importance: float = 0.0) -> None:
        self.handles = []
        self.metta_expression = {}
        if handle is not None:
            self.handles.append(handle)
        self.importance = importance
        self.strength = 0.0
        self.assignment = Assignment()

    @classmethod
    def copy(cls, other: "QueryAnswer") -> "QueryAnswer":
        """
        Creates a deep copy of another QueryAnswer.

        Args:
            other (QueryAnswer): The QueryAnswer to copy.

        Returns:
            QueryAnswer: A new instance with duplicated data.
        """
        q = cls(importance=other.importance)
        q.strength = other.strength
        q.handles = other.handles
        q.assignment.copy_from(other.assignment)
        return q

    def add_handle(self, handle: str) -> None:
        """
        Append a new handle to the list.
        """
        self.handles.append(handle)

    def merge(self, other: "QueryAnswer", merge_handles: bool = True) -> bool:
        """
        Merges another QueryAnswer into this one if assignments are compatible.

        Args:
            other (QueryAnswer): The other QueryAnswer to merge.
            merge_handles (bool): Whether to merge handle lists and importance.

        Returns:
            bool: True if merge succeeded, False if assignments were incompatible.
        """
        if not self.assignment.is_compatible(other.assignment):
            return False

        self.assignment.add_assignments(other.assignment)

        if merge_handles:
            self.importance = max(self.importance, other.importance)
            self.strength = self.strength * other.strength
            existing = set(self.handles)
            for handle in other.handles:
                if handle not in existing:
                    self.handles.append(handle)
                    existing.add(handle)

        return True

    def tokenize(self) -> str:
        """
        Produces a space-delimited token string representation.

        Format: "importance handles_count handles... assignment_count label value...".

        Returns:
            str: Tokenized string encoding this QueryAnswer.
        """
        tokens = [f"{self.strength:.10f} {self.importance:.10f}"]
        tokens.append(str(len(self.handles)))
        tokens.extend(self.handles)
        tokens.append(str(self.assignment.variable_count()))
        for label, value in self.assignment._mapping.items():
            tokens.append(label)
            tokens.append(value)
        return " ".join(tokens)

    def untokenize(self, token_str: str) -> None:
        """
        Parses a token string to populate this QueryAnswer's fields.

        Args:
            token_str (str): String produced by `tokenize()`.

        Raises:
            ValueError: If the token string format is invalid.
        """
        log.debug(f"Untokenizing QueryAnswer from: {token_str}")
        tokens = token_str.strip().split()
        cursor = 0

        def next_token():
            nonlocal cursor
            if cursor >= len(tokens):
                raise ValueError("Invalid token string: unexpected end of tokens")
            token = tokens[cursor]
            cursor += 1
            return token

        try:
            self.strength = float(next_token())
            self.importance = float(next_token())

            handles_size = int(next_token())
            if handles_size < 0:
                raise ValueError(f"Handles size cannot be negative: {handles_size}")

            self.handles = [next_token() for _ in range(handles_size)]

            assignment_size = int(next_token())
            if assignment_size < 0:
                raise ValueError(f"Assignment size cannot be negative: {assignment_size}")

            self.assignment = Assignment()
            for _ in range(assignment_size):
                label = next_token()
                value = next_token()
                self.assignment.assign(label, value)
                log.debug(f"Parsed assignment: ({label}: {value})")

            metta_expression_size = int(next_token())

            for _ in range(metta_expression_size + 1):
                key = next_token()
                self.metta_expression[key] = next_token()

            log.debug(f"Metta expression: {self.metta_expression}")

            if cursor != len(tokens):
                log.error(f"cursor: {cursor}, tokens: {tokens}")
                raise ValueError("Invalid token string: extra data after parsing")

            log.debug(f"QueryAnswer untokenized successfully: {self.to_string()}")

        except (ValueError, IndexError) as e:
            log.error(str(e))
            raise

    def to_string(self) -> str:
        """
        Returns a human-readable string representation of the QueryAnswer.

        Format:
            QueryAnswer<handles_count,assignment_count> [handles...] {assignments...} importance
        """
        handles_str = ", ".join(self.handles)

        return f"QueryAnswer<{len(self.handles)},{self.assignment.variable_count()}> [{handles_str}] {self.assignment.to_string()} ({self.strength:.6f}, {self.importance:.6f})"
